src/ui/public/test_harness/test_sharding/root_describe_call_filter.js

```diff
@@ -0,0 +1,81 @@
+/**
+ *  Intercept all calls to mocha.describe() and determine
+ *  which calls make it through using a filter function.
+ *
+ *  The filter function is also only called for ROOT describe()
+ *  calls, and all describe calls inside any allowed call are
+ *  automatically passed through.
+ *
+ *  ## example
+ *
+ *  assume tests that look like this:
+ *
+ *  ```js
+ *  describe('section 1', () => {
+ *    describe('item 1', () => {
+ *
+ *    })
+ *  })
+ *  ```
+ *
+ *  If the filter function returned true for "section 1" then "item 1"
+ *  would automatically be authorized. If it returned false for "section 1"
+ *  then "section 1" would be ignored and "item 1" would never be defined
+ *
+ *  @param {function} test - a function that takes the first argument
+ *                             passed to describe, the sections name
+ *  @return {undefined}
+ */
+export function setupRootDescribeCallFilter(test) {
+  const originalDescribe = window.describe;
+
+  if (!originalDescribe) {
+    throw new TypeError('window.describe must be defined by mocha before test sharding can be setup');
+  }
+
+  /**
+   *  In order to track which describe calls are nested
+   *  within other calls, we track our current level as a
+   *  number. Before passing any describe call to mocha we
+   *  increment that counter, and once mocha is done we
+   *  decrement it. That way, when we are at the root level
+   *  our "depth" will equal 0
```

Comment:
How about this?

"When `describe` is called, it will in turn call additional nested `describes`. We'll track how many nested `describe` calls have been made by incrementing `describeCallDepth` each time `describe` is called until the leaf `describe` call has been reached. To allow `describeCallDepth` to reset to 0, we'll decrement it when the nested `describe` calls for any given `describe` have completed."